python/strings_udf/cpp/src/strings/udf/shim.cu

```diff
@@ -223,7 +223,7 @@ extern "C" __device__ int udf_string_from_string_view(int* nb_retbal,
                                                       void* udf_str)
 {
   auto str_view_ptr = reinterpret_cast<cudf::string_view const*>(str);
-  auto udf_str_ptr  = reinterpret_cast<udf_string*>(udf_str);
+  auto udf_str_ptr  = new (udf_str) udf_string;
   *udf_str_ptr      = udf_string(*str_view_ptr);
 
   return 0;
```

Comment:
I think I don't understand what is going on here (and, _mutatis mutandis_, the rest of the patch).

`udf_string{string_view}` constructs a new `udf_string` from a `string_view` by calling `udf_string{sv.data(), sv.size_bytes()}` which sets up `m_bytes`, `m_capacity` with `size_bytes`, does `malloc` of the capacity and then a memcpy of the data into the `m_data` field.

So we did a malloc + memcpy to make the rvalue on line 227.

Separately we use placement new to build a pointer to some heap allocated space which default-initialises `udf_str_ptr` to be (say) zero-initialised.

Finally, the assignment `*udf_str_ptr = udf_string(*str_view_ptr)` calls `operator=` which does a move-construction (via `udf_string::assign` ?).

Why this level of indirection, why not:
```
auto udf_str_ptr = new (udf_str) udf_string;
(*udf_str_ptr)::assign(*reinterpret_cast<string_view *>(str));
```

What am I missing?

Reply:
That describes the process as I understand it as well, is your comment more about this being too verbose or is there something off about what I'm intending to happen here?

Reply:
No sorry, I got confused by all the `operator=` overloading. I was worried that there was an extra copy happening, but I realise no.